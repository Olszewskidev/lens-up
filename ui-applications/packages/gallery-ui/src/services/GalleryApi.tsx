import { HubConnectionBuilder } from '@microsoft/signalr';
import { useCypressSignalRMock as MockHubConnectionBuild, hubPublish } from 'cypress-signalr-mock';
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { LoginToGalleryPayload, LoginToGalleryResponse, PhotoItem } from '../types/GalleryApiTypes';

const baseQuery = fetchBaseQuery({
    baseUrl: import.meta.env.VITE_GALLERY_SERVICE_URL,
});

export const galleryApi = createApi({
    reducerPath: 'galleryApi',
    baseQuery: baseQuery,
    endpoints: (builder) => ({
        loginToGallery: builder.mutation<LoginToGalleryResponse, LoginToGalleryPayload>({
            query: (payload) => ({
                method: 'POST',
                url: 'login',
                body: payload,
            }),
        }),
        getGalleryPhotos: builder.query<PhotoItem[], string>({
            queryFn: () => ({ data: [] }),
            async onCacheEntryAdded(galleryId, { cacheDataLoaded, cacheEntryRemoved, updateCachedData },) {
                try {
                    await cacheDataLoaded;

                    const socket = MockHubConnectionBuild("testhub", {
                        enableForVitest: true,
                    }) ??
                        new HubConnectionBuilder().withUrl(`${import.meta.env.VITE_GALLERY_SERVICE_URL}/hubs/gallery?galleryId=${galleryId}`).build();

                    await socket.start().then(() => {
                        console.log("Connected to socket.")
                    }).catch(() => {
                        console.error("Failed during socket connection.")
                    })

                    socket.on("PhotoUploadedToGallery", (message: PhotoItem) => {
                        console.log("PhotoUploadedToGallery notification")
                        updateCachedData((draft) => {
                            return [...draft, message];
                        });
                    });
                    
                    await cacheEntryRemoved;

                    socket.off('connect');
                    socket.off("PhotoUploadedToGallery");
                } catch {
                    console.error("Failed during socket connection.")
                }
            }
        })
    }),
})

export const { useLoginToGalleryMutation, useGetGalleryPhotosQuery } = galleryApi