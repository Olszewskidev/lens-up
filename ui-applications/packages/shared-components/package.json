{
  "name": "@lens-up/shared-components",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "main": "./src/index.ts",
  "scripts": {
    "build": "tsc && vite build",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview"
  },
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@types/react": "^18.2.66",
    "@types/react-dom": "^18.2.22",
    "@typescript-eslint/eslint-plugin": "^7.2.0",
    "@typescript-eslint/parser": "^7.2.0",
    "@vitejs/plugin-react-swc": "^3.5.0",
    "eslint": "^8.57.0",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-react-refresh": "^0.4.6",
    "typescript": "^5.2.2",
    "vite": "^5.2.0"
  },
  "jest": {
    "moduleNameMapper": {
         "\\.(jpg|ico|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/mocks/fileMock.js",
         "\\.(css|less)$": "<rootDir>/mocks/fileMock.js"
       }
   }
}
