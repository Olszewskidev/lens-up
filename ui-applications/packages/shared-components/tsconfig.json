{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "module": "es2020",
    "target": "es2020",
    "moduleResolution": "bundler"
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"],
  "references": [{ "path": "./tsconfig.node.json" }],
  "jest": {
    "moduleNameMapper": {
         "\\.(jpg|ico|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/mocks/fileMock.js",
         "\\.(css|less)$": "<rootDir>/mocks/fileMock.js"
       }
   }
}