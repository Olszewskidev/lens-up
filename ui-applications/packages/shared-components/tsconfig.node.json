{
  "compilerOptions": {
    "target": "es2020",
    "composite": true,
    "skipLibCheck": true,
    "module": "es2020",
    "moduleResolution": "bundler",
    "allowSyntheticDefaultImports": true,
    "strict": true
  },
  "include": ["vite.config.ts"]
}
