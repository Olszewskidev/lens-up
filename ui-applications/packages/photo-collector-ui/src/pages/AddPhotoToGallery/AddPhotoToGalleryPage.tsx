import { ChangeEvent, useState } from "react";
import { useAddPhotoToGalleryMutation } from "../../services/PhotoCollectorApi";
import { AddPhotoForm, AddPhotoFormSkeleton } from "./components"
import { ErrorCardComponent, SuccessAlertComponent } from "@lens-up/shared-components";
import { useParams } from "react-router-dom";

const AddPhotoToGalleryPage = () => {
    const { enterCode } = useParams();
    const [addPhoto, { isError, isSuccess, isUninitialized, isLoading }] = useAddPhotoToGalleryMutation();
    const [photo, setPhoto] = useState<File>();

    const handlePhotoInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        // TODO: Validation
        const fileList = event.target.files
        const isEmpty = !fileList || fileList.length === 0;
        if (isEmpty) {
            return
        }
        const photo = fileList[0];
        setPhoto(photo)
    };

    const handleFormSubmit = () => {
        // TODO: Validation
        if (!photo || !enterCode) {
            return
        }
        const formData = new FormData();
        formData.append('File', photo);
        addPhoto({ enterCode, formData })
    }

    return (
        <>
            {
                isUninitialized && <AddPhotoForm
                    handleFormSubmit={handleFormSubmit}
                    handlePhotoInputChange={handlePhotoInputChange}
                    photo={photo}
                    isLoading={isLoading} />
            }
            {
                isError && <ErrorCardComponent />
            }
            {
                isSuccess && <SuccessAlertComponent />
            }
        </>
    )
}

export default AddPhotoToGalleryPage
