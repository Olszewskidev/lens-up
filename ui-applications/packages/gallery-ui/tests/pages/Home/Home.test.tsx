import { buildQueries, cleanup, fireEvent, getByAltText, getSuggestedQuery, render, screen, waitFor } from '@testing-library/react'
import { expect, test, describe, vi, beforeAll, beforeEach, afterEach, afterAll } from 'vitest';
import { HubConnectionBuilder } from '@microsoft/signalr';
import { setupServer } from 'msw/node';
import { loginSubmit } from '../../utils/LoginEvents.tsx';
import { getQRCodeUrl } from '../../../src/utils/qRCodeHelper.ts';
import { HttpResponse, http as mswhttp } from 'msw';
import { LoginToGalleryResponse } from '../../../src/types/GalleryApiTypes.ts';
import { Server, Socket } from "socket.io";
import { io as ioc, Socket as ClientSocket } from "socket.io-client";
import { IncomingMessage, ServerResponse } from "http";
import http from "http";
import { AddressInfo } from 'node:net';
import { startServer } from "../../webSocketTestUtils.tsx";
import { store } from '../../../src/app/store/store.ts';
import { store as photoStore } from '../../../../photo-collector-ui/src/app/store.ts';
import { Provider } from 'react-redux'
import { BrowserRouter, Navigate, Route, Routes } from 'react-router-dom';
import AddPhotoToGalleryPage from "../../../../photo-collector-ui/src/pages/AddPhotoToGallery/AddPhotoToGalleryPage.tsx";
import { ErrorCardComponent, SuccessCardComponent } from '@lens-up/shared-components';
import { HomePage, LoginPage } from '../../../src/pages/index.ts';
import { galleryApi, useGetGalleryPhotosQuery } from '../../../src/services/GalleryApi.tsx';

const photo = new File(["happy"], "happy.png", { type: "image/png" });

function waitForSocket(socket: Socket | ClientSocket, event: string) {
  return new Promise((resolve) => {
    socket.once(event, resolve);
  });
}

let handlers = [
  mswhttp.post(`${import.meta.env.VITE_GALLERY_SERVICE_URL}/login`, async () => {
    let response: LoginToGalleryResponse = { enterCode: '0', galleryId: '0', qrCodeUrl: 'QRCodeUrl' };
    let res = await HttpResponse.json(response);
    return res;
  }),
  mswhttp.post(`${import.meta.env.VITE_PHOTO_COLLECTOR_SERVICE_URL}/upload-photo/0`, async () => {
    const socket = new HubConnectionBuilder().withUrl(`${import.meta.env.VITE_GALLERY_SERVICE_URL}/hubs/gallery?galleryId=0`).build();

    await socket.start().then(() => {
      console.log("Connected to socket.")
    }).catch(() => {
      console.error("Failed during socket connection.")
    })

    socket.send("PhotoUploadedToGallery", photo);

    return new Response(null, {
      status: 200,
      headers: {
        Allow: 'GET,HEAD,POST',
      },
    })
  })
]

let asFragment: () => DocumentFragment;
let HomeasFragment: () => DocumentFragment;
let baseElement: HTMLElement;
let homeRender: DocumentFragment;

/**
* @vitest-environment jsdom
*/
describe("Home page with no photos", async () => {
  const server = setupServer(...handlers);

  beforeAll(async () => {
    server.listen();
    ({ asFragment, baseElement } = await loginSubmit());
  })
  beforeEach(async () => waitFor(() => {
    expect(global.window.location.pathname).contains("/gallery/0");
    homeRender = asFragment();
  }))
  afterEach(cleanup);
  afterAll(() => {
    server.close()
  })

  test("No photo hub must show qr code card in home page", async () => {
    const firstRender = asFragment();

    expect(firstRender).toMatchSnapshot();
  });

  test("No photo hub must have qr code url in image source", async () => {
    const firstRender = asFragment();

    expect(getByAltText(baseElement, "QR code").getAttribute("src")).toEqual(getQRCodeUrl());

    expect(firstRender).toMatchSnapshot();
  });
});

const successCardRender = render(
  <SuccessCardComponent title="Congratulations!" text="You just joined to the party." />
).asFragment();

describe("Home page with photos", async () => {
  const server = setupServer(...handlers);

  beforeAll(async () => {

    server.listen();
    //({ asFragment, baseElement } = await loginSubmit());
    //HomeasFragment = asFragment;
    console.log(global.window.location.href);
  })
  beforeEach(async () => waitFor(() => {
    //expect(global.window.location.pathname).contains("/gallery/0");
    //homeRender = asFragment();
  }))
  afterEach(cleanup);
  afterAll(() => {
    server.close()
  })

  test("Photo gallery must be shown in home", async () => {
    URL.createObjectURL = vi.fn();
    const { container, baseElement, asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="/gallery/:enterCode" element={<HomePage />} />
          </Routes>
          <Navigate to="/gallery/0" replace={true} />
        </BrowserRouter>
      </Provider>
    );

    const hub = (new HubConnectionBuilder()).build();

    hub.invoke("PhotoUploadedToGallery", photo);
    hub.send("PhotoUploadedToGallery", photo);

    expect(asFragment()).toMatchSnapshot();
  });

  test("Photo gallery must be shown", async () => {
    console.log(global.window.location.href);
    URL.createObjectURL = vi.fn();
    const { container, baseElement, asFragment } = render(

      <Provider store={photoStore}>
        <BrowserRouter>
          <Routes>
            <Route path="/upload-photo/:enterCode" element={<AddPhotoToGalleryPage />} />
          </Routes>
          <Navigate to="/upload-photo/0" replace={true} />
        </BrowserRouter>
      </Provider>

    );

    const element = container.querySelector("#dropzone-file") as HTMLElement;

    console.log(global.window.location.href);

    fireEvent.change(element, { target: { files: [photo] } });

    fireEvent.change(screen.getByPlaceholderText("Your name"), { target: { value: "author" } });

    fireEvent.change(screen.getByPlaceholderText("..."), { target: { value: "happy" } });

    fireEvent.submit(screen.getByRole('button'));

    let successRender = asFragment();

    await waitFor(() => {
      successRender = asFragment();
      expect(successRender).toEqual(successCardRender);
    }, { timeout: 50000 }).then(() => expect(asFragment()).toMatchSnapshot());
  }, { timeout: 50000 });
})