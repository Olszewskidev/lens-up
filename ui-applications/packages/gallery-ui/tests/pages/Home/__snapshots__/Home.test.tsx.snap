// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page with photos > Photo gallery must be shown in home 1`] = `
<DocumentFragment>
  <div
    class="absolute top-1/2 left-1/2 transform -translate-x-1/2"
  >
    <figure>
      <img
        alt="QR code"
        class="rounded-lg w-64 h-64"
        src="QRCodeUrl"
      />
      <figcaption
        class="font-bold text-base mb-2 text-center text-white"
      >
        Scan this and join.
      </figcaption>
    </figure>
  </div>
</DocumentFragment>
`;
